utils/s2n_random.c

```diff
@@ -553,6 +553,10 @@ S2N_RESULT s2n_rand_init(void)
         return S2N_RESULT_OK;
     }
 
+    /* Unset any existing random engine */
+    RAND_set_rand_engine(NULL);
+    RAND_set_rand_method(NULL);
+
```

Comment:
Why do these methods work, but how we set the engine / method below don't work? Should we be using these methods to set our randomness instead?

Reply:
It's due to how Openssl gets a random method to use when it needs random bytes. Basically there is a global variable called[ default_rand_meth which is used over any ](https://github.com/openssl/openssl/blob/openssl-3.0/crypto/rand/rand_lib.c#L193)random engine. Normally, we are able to set our random engine as the default. However, if you call RAND_bytes before setting up our engine, that default_rand_method gets set to the Openssl randomness function, and therefore our randomness method will never be chosen over Openssl's.
To get around this we have to call a function that alters the default_rand_method variable after it's already been set. So we basically have two options, either clean up the default random method before setting up the engine code(what I did), or calling RAND_set_rand_engine() with our custom rand method. I don't think it really matters which one we choose.

Reply:
Another piece of the puzzle, apparently we [used to use the RAND_set_rand_method()](https://github.com/aws/s2n-tls/commit/19f7a5dce7b81f0f32522176c1a8b8e023daa402) before replacing it with ENGINE code. But it's very weird that the ENGINE code doesn't fully replace the RAND code.

Reply:
The behavior of the `default_rand_meth` global is very confusing. If I'm reading the OpenSSL documentation correctly, this shouldn't be happening:
https://www.openssl.org/docs/man1.0.2/man3/RAND_set_rand_method.html#notes
> If a default ENGINE is specified for RAND functionality using an ENGINE API function, that will override any RAND defaults set using the RAND API (ie. RAND_set_rand_method()). For this reason, the ENGINE API is the recommended way to control default implementations for use in RAND and other cryptographic algorithms.

But other than calling `RAND_set_rand_method()/RAND_set_rand_engine()`, there doesn't seem to be a way to override the `default_rand_meth` variable after it's set, except for `OPENSSL_cleanup()`.

Given that the RAND methods are deprecated, it seems like it might be best to use Madeleine's current implementation and keep the ENGINE APIs to actually set our random implementation.